File: node.js/faults/redis-fault.js
const ChaosFault = require('./fault');
const Config = require('../config');

class RedisLatencyFault extends ChaosFault {
    constructor(redisClient) {
        super();
        this.active = false;
        this.redisClient = redisClient;
        this.delay = 0;
    }

    getName() {
        return 'redis_latency';
    }

    async start(params = {}) {
        const config = Config.loadConfig();
        let delay = config.faults.redis.defaultDelay;

        if (params.duration) {
            delay = parseInt(params.duration);
        }

        if (this.active && this.delay == delay) {
            return;
        }

        try {
            if (this.redisClient && this.redisClient.startFault) {
                await this.redisClient.startFault(delay);
                this.active = true;
                this.delay = delay;
            } else {
                this.active = true;
            }
        } catch (error) {
            throw error;
        }
    }

    async stop() {
        if (!this.active) {
            return;
        }

        try {
            if (this.redisClient && this.redisClient.stopFault) {
                await this.redisClient.stopFault();
            }
            this.active = false;
        } catch (error) {
            throw error;
        }
    }

    isActive() {
        return this.active;
    }
}

module.exports = RedisLatencyFault;

File: python/faults/redis_fault.py
# python/faults/redis_fault.py
import logging
import threading

from faults.fault_manager import ChaosFault
from config import load_config
from storage import RedisClient

logger = logging.getLogger(__name__)

class RedisLatencyFault(ChaosFault):
    def __init__(self, redis_client: RedisClient):
        self._active = False
        self._lock = threading.Lock()
        self._redis_client = redis_client
        self._original_redis_get = None
        self._original_redis_set = None
        self._fault_delay = 0

    def name(self) -> str:
        return "redis_latency"

    def is_active(self) -> bool:
        with self._lock:
            return self._active

    def start(self, params: dict) -> None:
        with self._lock:
            if self._active:
                logger.info("Redis latency fault already active.")
                return

            config = load_config()
            delay_ms = params.get("duration", config.faults.redis.default_delay)

            self._fault_delay = delay_ms

            if self._redis_client:
                self._redis_client.start_fault(delay_ms)

            self._active = True

    def stop(self) -> None:
        with self._lock:
            if not self._active:
                return

            if self._redis_client:
                self._redis_client.stop_fault()

            self._active = False
            self._fault_delay = 0
            logger.info("Redis latency fault stopped.")
